[C23] Do not diagnose binary literals as an extension

AaronBallman opened the pull request:
We previously would diagnose them as a GNU extension in C mode, but they are now a feature of C23. The -Wgnu-binary-literal warning group no longer controls any diagnostics as this is no longer a GNU extension. The warning group is retained as a noop to help avoid "unknown warning" diagnostics.

This also adds the companion compatibility warning which existed for C++ but not for C.

Fixes https://github.com/llvm/llvm-project/issues/72017

AaronBallman committed:
[C23] Do not diagnose binary literals as an extension

We previously would diagnose them as a GNU extension in C mode, but
they are now a feature of C23. This removes the old warning group,
-Wgnu-binary-literal, as it no longer makes sense and searching through
public repositories shows it is a rarely used option (same for the -Wno
form). It also adds the companion compatibility warning which existed
for C++ but not for C.

Fixes https://github.com/llvm/llvm-project/issues/72017

AaronBallman committed:
Fix incorrect indentation for sphinx

efriedma-quic reviewed (commented):
We could consider keeping the warning group, not actually guarding any warning, so we don't warn on `-Wno-gnu-binary-literal`.

Otherwise looks fine.

shafik reviewed (commented):
LGTM

AaronBallman committed:
Restore the -Wgnu-binary-literal group as a noop

This helps avoid "unknown warning group" diagnostics even though it
doesn't actually control a diagnostic.

AaronBallman commented:
> We could consider keeping the warning group, not actually guarding any warning, so we don't warn on `-Wno-gnu-binary-literal`.

That's a good idea, I've done that and updated the release notes accordingly.
